### PR TITLE
Groupable, supply limitable NFT tokens as derivative of fungible tokens

### PR DESCRIPTION
**Pull request by imaginaryusername**

### CODE REVIEW

**Comment by markblundeberg** [comment]
Interesting idea!

It may be useful if NFT GENESIS transactions would indicate which input contained the burned token.

I don't get what NFT2 scheme gets you ... why can't the minting authority just make NFT1 and keep all those tokens & baton on their own address?

**Comment by imaginaryusername** [comment]
>if NFT GENESIS transactions would indicate which input contained the burned token

Ah, in case there are multiple tokens burned? Good thought!

>just make NFT1 and keep all those tokens & baton on their own address?

That was my initial idea - the NFT2 modification is to make sure if any minting tokens every get to another party (mistake? theft?) they will be rendered invalid.

**Comment by Kronkmeister** [comment]
I wanted to create 7 Tokens each identifiable with a simple number from 1-7. I'd love to see it implemented! Thanks

**Comment by jcramer** [comment]
To continue the conversation from Telegram, what do you guys think about using an XPUB as a way to represent group NFTs?  Simply paste the XPUB into the token's Document URI field and then anyone can easily verify if the token belongs to the same XPUB group by looking at the signed inputs to the GENESIS transaction.

**Comment by jcramer** [comment]
> I wanted to create 7 Tokens each identifiable with a simple number from 1-7. I'd love to see it implemented! Thanks

To get a unique ID for your tokens application you have two options:

1. Use the GENESIS fields to manually add the token's ID (e.g., `Token Name = 1`)
2. Order the tokens by tokenID (least to greatest) to determine which number the token belongs to 1-7.  For example, tokenId `0f1a...0124` = 1, `3032...aa11` = 2, `3122...3155` = 3, etc.

**Comment by imaginaryusername** [comment]
>To continue the conversation from Telegram, what do you guys think about using an XPUB as a way to represent group NFTs? 

In the context of my proposal, using xpub has two downsides:

1. The number of tokens cannot be limited without additional specification.

2. Only owner of the xpriv can issue tokens.  (In first part of proposal, anyone who has a fungible token can burn it to generate NFT; fungible token can pass through many hands).

If token emission by strictly the same owner is desired, one might as well make it single address only (second part); and the proposal takes advantage of exiting specs for fungible token quantities (can be limited, can pass baton), both of which requires additional fields for xpub.

File: NFT.md [hunk 1]
```diff
@@ -48,13 +48,13 @@ As described above, a problem with the "one genesis per token" approach is that
...
+To forge an NFT token, first prepare an UTXO with one or more of the NFT-enabled fungible tokens associated with it; then spend that UTXO in a GENESIS transaction that generates a new token with quantity of 1 and decimal places 0, prefixing its document URL with "NFT1_" followed by token hash of the fungible token for quick lookup, followed by a two-byte number indicating which input contained the "generating" token in question. This simultaneously generates a new token and burn one fungible token. Wallets that support this scheme can check whether a given NFT token is valid by tracing DAG up to genesis, check if the tx burned any of the fungible token it specified, and check the validity of that back to the genesis of said fungible token. Each NFT token generated this way will have its own secondary "NFT id" from the txid of its genesis, and can remain distinct from each other.
```
Comment (jcramer) [inline NFT.md:51]: 1. suggest changing "; then spend that UTXO in a GENESIS transaction" to. "; then burn/spend that token UTXO in a GENESIS transaction"

Comment (jcramer) [reply NFT.md:51]: 2. why not just a 1-byte number indicating which input contained? Do you expect it will be greater than index 255?

Comment (imaginaryusername) [reply NFT.md:51]: 1. Good point! Should I make a follow up PR? You can also fix that at your convenience. 

2. Transactions can already have more than 255 inputs https://blockchair.com/bitcoin-cash/transaction/dd111ad72411041e2753382095dd41a580d33f489766100c5404d22c11807a1f ; while a two byte field is probably enough for all forseeable future, using the token ID is probably more robust.

**Comment by markblundeberg** [comment]
> To continue the conversation from Telegram, what do you guys think about using an XPUB as a way to represent group NFTs? Simply paste the XPUB into the token's Document URI field and then anyone can easily verify if the token belongs to the same XPUB group by looking at the signed inputs to the GENESIS transaction.

I was thinking about this a bit --- if you share an xpub with others, there is a [security downside that many people don't know about](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#implications) -- "knowledge of a parent extended public key plus any non-hardened private key descending from it is equivalent to knowing the parent extended private key (and thus every private and public key descending from it).". I.e., if you share xpub then accidentally leak *just one* private key then the whole xprv becomes compromised -- all private keys compromised.

In addition, if you publish an xpub then the reason for the existence of the xpub (privacy) is nullified. So you may as well not use an xpub and just issue from same single address.

It's also not so easy to verify if a key belongs to a given xpub, if you have to search through potentially tens of thousands of HD key derivations to find a match.

File: NFT.md (hunk 1 above)
Comment (jcramer) [reply NFT.md:51]: I think we can just leave it for now.  :) Thank you for this contribution, I think this idea is going to take NFTs to the next level!

In the Electron Cash we can add a method & UI to burn tokens.

After that's added then we can extend that burn method to allow the user to right-click on any token and select "Mint a NFT" from any fungible token.

**Comment by jcramer** [comment]
ok so @Kronkmeister probably not ideal to use XPUB to "group" tokens.  As you can see we just merged in @imaginaryusername 's idea here, so we should try to use this moving forward for grouping together NFTs.  SLPJS will need to be updated with helper methods to allow & validate for this.

**Pull request merged**